CORE-8496 k/server: improve full disk produce error response

pgellert opened the pull request:
This improves the client-side observability of the scenario when produce
requests are getting rejected due to brokers running out of disk space.

We change the error code to kafka_storage_error conditional on a
sufficiently up to date produce api version, and add a custom error
message to the response.

Note that some older clients that do not know about kafka_storage_error
would handle it as an unknown_error. While kafka_storage_error and
broker_not_available are retriable, unknown_server_error is not. A
retriable error is preferred for easier recovery once disk health
recovers. Hence, we condition the return on supporting at least api
version 4.

The kafka-client version is bumped to get support for the
KafkaStorageError.

Fixes https://redpandadata.atlassian.net/browse/CORE-8496

Relates to https://github.com/redpanda-data/redpanda/issues/24492

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x
- [ ] v24.1.x

## Release Notes
### Improvements
* Improve produce error response when produce requests are rejected due to full disk.
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert committed:
k/server: refactor produce make_error_response

Pure refactor to allow setting an optional error message in the
response.

pgellert committed:
k/server: improve full disk produce error response

This improves the client-side observability of the scenario when produce
requests are getting rejected due to brokers running out of disk space.

We change the error code to kafka_storage_error conditional on a
sufficiently up to date produce api version, and add a custom error
message to the response.

Note that some older clients that do not know about kafka_storage_error
would handle it as an unknown_error. While kafka_storage_error and
broker_not_available are retriable, unknown_server_error is not. A
retriable error is preferred for easier recovery once disk health
recovers. Hence, we condition the return on supporting at least api
version 4.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64116-->
<!--2dcc83d3-33a7-44f8-bcd4-a4b5faf64b02-->
<!--bf77295f-cf17-418f-90e0-de2b1eb898ae-->
<!--title-->
## Retry command for [Build#64116](https://buildkite.com/redpanda/redpanda/builds/64116)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/compatibility/java_compression_test.py::JavaCompressionTest.test_upgrade_java_compression@{"compression_type":"snappy"}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d2edc09c-07d0-4567-9a74-7b5705a45c56-->
<!--TEST RESULTS-->
<!--492031be-ff0f-4edc-a40f-af6474e7417c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64116">build#64116</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/64116#0195f737-e071-4fd8-aa9f-737f02815faf | FLAKY | 1/2 |
| rptest.tests.compatibility.java_compression_test.JavaCompressionTest.test_upgrade_java_compression.compression_type=CompressionTypes.SNAPPY | ducktape | https://buildkite.com/redpanda/redpanda/builds/64116#0195f734-34d9-4c2f-8169-9faf90bf7e34 | FAIL | 0/20 |
| rptest.tests.compatibility.java_compression_test.JavaCompressionTest.test_upgrade_java_compression.compression_type=CompressionTypes.SNAPPY | ducktape | https://buildkite.com/redpanda/redpanda/builds/64116#0195f737-e06f-4e73-88b7-36c131c7881d | FAIL | 0/20 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64116#0195f734-34d9-4c2f-8169-9faf90bf7e34 | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/64116#0195f734-34d9-4753-96c0-3c5a6489ed7f | FLAKY | 1/2 |
| rptest.tests.leadership_transfer_test.LeadershipPinningTest.test_leadership_pinning_sanctions | ducktape | https://buildkite.com/redpanda/redpanda/builds/64116#0195f734-34d9-4753-96c0-3c5a6489ed7f | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64116#0195f737-e070-42fc-a220-0ebdde7eb357 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64116#0195f737-e070-42fc-a220-0ebdde7eb357 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64116#0195f737-e070-42fc-a220-0ebdde7eb357 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64156">build#64156</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64156#0195fb5a-6d4c-46d8-9ac1-ae9b5d3e685e | FLAKY         | 1/2      |
| rptest.tests.datalake.rest_catalog_connection_test.RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog.cloud_storage_type=CloudStorageType.S3                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64156#0195fb5a-6d4b-4b89-9897-0244cb80a381 | FLAKY         | 1/2      |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64156#0195fb74-9278-4285-a5a1-d56da90f18d0 | FLAKY         | 1/2      |
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=start_offset                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64156#0195fb5a-6d4b-4853-ba04-cb47a3404ee0 | FLAKY         | 1/2      |
</details>
<!--/body-->

pgellert committed:
dt: bump kafka-python to 2.0.6

The version is updates to get support for KafkaStorageError used in the
next commit.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/95b6fc21a99544850d5d4d19eb3e1aa4e3a495f9..598e47da5d2183cfcee37eb97a2f0279a5197328): fix failing `JavaCompressionTest` after kafka-python version update

pgellert commented on tests/rptest/tests/compatibility/java_compression_test.py:
```diff
@@ -166,17 +166,7 @@ def test_upgrade_java_compression(self, compression_type):
         self.wait_for_compacted_segments(expected_num_segments - 1)
 
-        if compression_type == TopicSpec.CompressionTypes.SNAPPY:
-            # Refer to the Github issue in the function comment above re: the issue with
```
Thanks for the great write-up here @WillemKauf!

WillemKauf commented:
> with franz/go you just hang....(will timeout) (https://github.com/redpanda-data/redpanda/issues/24492)

Does the added error message get printed in between retries instead of just appearing as a hang with franz/go?

michael-redpanda commented:
FYI: https://github.com/apache/kafka/blob/f1bb29b93a2a0d74cf393b7d6a7fa80b3945c105/clients/src/main/resources/common/message/ProduceResponse.json#L29

To go along with @pgellert's cover letter, here's the reference to apache kafka where produce response v4 adds suport for `kafka_storage_error`

michael-redpanda reviewed (approved):
🔥

michael-redpanda commented:
@pgellert - thoughts on backporting this?  I think the handling code is good to backport, but not sure about the dependency updates for testing

pgellert commented:
> > with franz/go you just hang....(will timeout) (#24492)
> 
> Does the added error message get printed in between retries instead of just appearing as a hang with franz/go?

Unfortunately not. I've only seen the java client surface the error message so far. Other clients like franz-go only care about the error code. This is why changing the error code to something that suggests "disk-related" should help here and unfortunately the most we can do I believe.

pgellert commented:
> thoughts on backporting this? I think the handling code is good to backport, but not sure about the dependency updates for testing

Technically no reason not to backport. I think even the test dependency update would be backportable (except for maybe the change to the ducktape test, but that probably doesn't exist in earlier branches anyways).

I went with no backport because I considered this an improvement, not a bug fix. But I am happy to backport if you think otherwise.

michael-redpanda commented:
> I went with no backport because I considered this an improvement, not a bug fix. But I am happy to backport if you think otherwise.

My feeling is that this could be considered a bug fix as we weren't actually honoring v4 of Produce API.

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25698-v24.3.x-717 remotes/upstream/v24.3.x
git cherry-pick -x 30ddb8c259 4954512794 598e47da5d
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14261739131).

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25698-v24.2.x-930 remotes/upstream/v24.2.x
git cherry-pick -x 30ddb8c259 4954512794 598e47da5d
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14261739966).
